django_socio_grpc/servicer_proxy.py

```diff
@@ -14,117 +21,175 @@
 )
 from django_socio_grpc.settings import grpc_settings
 
+if TYPE_CHECKING:
+    from django_socio_grpc.services import Service
+
 logger = logging.getLogger("django_socio_grpc")
 
 
+@dataclass
+class GRPCRequestContainer:
+    message: Message
+    context: GRPCSocioProxyContext
+    action: str
+    proxy: "ServicerProxy"
+
+
 class ServicerProxy:
-    def __init__(self, ServiceClass, **initkwargs):
-        self.service_class = ServiceClass
+    def __init__(self, service_class: Type["Service"], **initkwargs):
+        self.service_class = service_class
         self.initkwargs = initkwargs
 
-    def call_handler(self, action):
+        self.load_middleware(is_async=grpc_settings.GRPC_ASYNC)
+
+    def load_middleware(self, is_async=False):
+        """
+        Populate middleware lists from settings.MIDDLEWARE.
+        Must be called after the environment is fixed (see __call__ in subclasses).
+        From django BaseHandler.load_middleware
+        """
+        self._exception_middleware = []
+
+        get_response = self._get_response_async if is_async else self._get_response
+        handler = convert_exception_to_response(get_response)
+        handler_is_async = is_async
+        for middleware_path in reversed(grpc_settings.GRPC_MIDDLEWARE):
+            middleware = import_string(middleware_path)
+            middleware_can_sync = getattr(middleware, "sync_capable", True)
+            middleware_can_async = getattr(middleware, "async_capable", False)
+            if not middleware_can_sync and not middleware_can_async:
+                raise RuntimeError(
+                    "Middleware %s must have at least one of "
+                    "sync_capable/async_capable set to True." % middleware_path
+                )
+            elif not handler_is_async and middleware_can_sync:
+                middleware_is_async = False
+            else:
+                middleware_is_async = middleware_can_async
+            try:
+                # Adapt handler, if needed.
```

Comment:
Please follow comment convention.
+ What do the adaptation ? when this is needed ?

Reply:
All this code is an adaptation of Django BaseHandler.load_middleware with a few tweaks